src/snowplow/emitter.lua

```diff
@@ -0,0 +1,157 @@
+-- Copyright:   Copyright (c) 2022 Snowplow Analytics Ltd
+-- License:     Apache License Version 2.0
+
+local curl = require("cURL")
+local ss = require("lib.utils").safe_string -- Alias
+local set = require("lib.set")
+local validate = require("validate")
+
+local Emitter = {}
+local emitter = {}
+
+local VALID_PROTOCOLS = set.new_set({ "http", "https" })
+
+-- --------------------------------------------------------------
+-- Private methods
+
+-- Helper to generate the collector url from a given url.
+-- @param url string: The url of the collector
+-- @return string: The full collector URL
+local function as_collector_url(url, request_type)
+  local path = ({
+    ["GET"] = "/i",
+    ["POST"] = "/com.snowplowanalytics.snowplow/tp2",
+  })[request_type]
+
+  if url:find("://") then
+    local _, _, protocol = url:find("^(%a[^:]*)")
+    validate.is_string_from_set(VALID_PROTOCOLS, "protocol", protocol)
+    return url .. path
+  end
+
+  return "https://" .. url .. path
+end
+
+-- Decides if a given status code is considered a success.
+-- @param status_code number: The HTTP status code
+-- @return boolean, string: Whether event was successfully collected; and the reason for failure if not
+local function handle_status_code(status_code)
+  local code = tonumber(status_code)
+  if code == nil or code ~= math.floor(code) or code < 0 or code >= 600 then
+    return false, "Unrecognised status code [" .. ss(status_code) .. "]"
+  elseif code >= 400 and code < 500 then
+    return false, "HTTP status code [" .. ss(status_code) .. "] is a client error"
+  elseif code >= 500 then
+    return false, "HTTP status code [" .. ss(status_code) .. "] is a server error"
+  end
+
+  return true
+end
+
+-- Performs a GET request with a given payload to the collector.
+-- @param uri string: The URI (including querystring) to GET
+-- @return boolean, string: Whether event was successfully collected; and the reason for failure if not
+local function _get(self, payload)
+  local url = self.collector_url .. payload
+  -- `resp` is the table `:getinfo` reads from
+  local resp = {}
+  local c = curl.easy({
+    url = url,
+  }):setopt_writefunction(table.insert, resp)
+
+  local _, err = pcall(function()
+    c:perform()
+  end)
+
+  if err ~= nil then
+    return false, "Host [" .. self.collector_url .. payload .. "] not found (possible connectivity error)"
+  end
+
+  return handle_status_code(c:getinfo(curl.INFO_RESPONSE_CODE))
+end
+
+-- Performs a POST request with a given payload to the collector.
+-- @param uri string: The URI (including querystring) to GET
+-- @param method string: The HTTP method to use
+-- @return boolean, string: Whether event was successfully collected; and the reason for failure if not
+local function _post(self, payload)
+  -- `resp` is the table `:getinfo` reads from
+  local resp = {}
+  local c = curl.easy({
+    url = self.collector_url,
+    post = true,
+    httpheader = {
+      "Content-Type: application/json",
```

Comment:
It might not be necessary, but looking at other trackers, I see that they add the charset to the content-type: `application/json; charset=utf-8`. Do you think it's worth adding here?

Reply:
I can't see any harm in it! - It has been added.